File: code/siwonblue/week3/16434.js
const filePath =
  process.platform === "linux"
    ? "/dev/stdin"
    : "/Users/jeonsiwon/Desktop/Almut-3rd/code/siwonblue/week3/input.txt";
const input = require("fs")
  .readFileSync(filePath, "utf-8")
  .toString()
  .trim()
  .split("\n");
let [N, attack] = input[0].split(" ").map(Number);
const me = { max: -1, cur: -1, attack };
const maxHp = 10 ** 12 * 123456;

function getRoomInfo() {
  const rooms = [];
  for (let i = 1; i <= N; i++) {
    const [type, first, second] = [...input][i].split(" ").map(Number);
    const roomsInfo = { type };
    if (type === 1) {
      roomsInfo["attack"] = first;
      roomsInfo["stemina"] = second;
    } else {
      roomsInfo["buff"] = first;
      roomsInfo["potion"] = second;
    }
    rooms.push(roomsInfo);
  }
  return rooms;
}

function train(rooms) {
  let killDragon = true;
  for (let i = 0; i < rooms.length; i++) {
    const room = rooms[i];
    const { type } = room;
    // console.log("me", me);
    // console.log("room", room);
    if (type === 1) {
      let fightCount = 0;
      // fight simulation
      while (true) {
        fightCount += 1;
        room.stemina -= me.attack;
        if (room.stemina <= 0) break;

        me.cur -= room.attack;
        if (me.cur <= 0) {
          // console.log(`die in room ${i + 1}`);
          return { killDragon: false };
        }
      }
      // console.log(`${fightCount} fight and win`);
    } else {
      const { buff, potion } = room;
      me.attack += buff;
      me.cur += potion;
      if (me.cur >= me.max) me.cur = me.max;
    }

    // console.log(`room ${i + 1} ends and my condition is `, me);
    // console.log(`\n`);
  }
  return { killDragon };
}

function solution() {
  let start = 1;
  let end = maxHp;
  let ans = 10 ** 12;

  while (start <= end) {
    let mid = parseInt((start + end) / 2);
    me.cur = mid;
    me.max = mid;
    me.attack = attack;
    const rooms = getRoomInfo();

    const { killDragon } = train(rooms);

    if (killDragon) {
      // console.log("mid", mid);
      // console.log("me:", me);
      ans = Math.min(ans, mid);
      end = mid - 1;
    } else start = mid + 1;
  }
  return ans;
}

console.log(solution());

File: code/siwonblue/week3/16434.cpp
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

struct Me{
  long long cur;
  long long max;
  long long attack;
};

bool train(Me me, vector<vector<long long>> room_info){
  
  bool killDragon = true;
  for(long long i=0;i<room_info.size();i++){
    vector<long long> room = room_info[i];
    long long type = room[0];
    if(type==2){
      me.attack += room[1];
      me.cur += room[2];
      if(me.cur>=me.max) me.cur = me.max;
      continue;
    }
    // timeout code
    // while(true){
    //   room[2] -= me.attack;
    //   if(room[2]<=0) break;

    //   me.cur -= room[1];
    //   if(me.cur<=0){
    //     killDragon = false;
    //     return killDragon;
    //   }
    // }
    // fix
    long long t;
    if(room[2] % me.attack == 0) t = room[2]/me.attack - 1;
    else t = room[2]/me.attack;
    me.cur -= room[1] * t;
    if(me.cur <= 0) return false;
  }
  
  return killDragon;
}

int main(){
  ios_base::sync_with_stdio(false);
  cin.tie(0);

  // input
  long long N, attack;
  cin >> N >> attack;
  vector<vector<long long>> room_info(N, vector<long long>(3, 0));
  
  for(long long i=0;i<N;i++){
    long long type, a,b;
    cin >> type >> a >> b;
    room_info[i] = vector<long long>{type,a,b};
  };

  long long maxHp = 1e18;

  Me me = {-1, -1, attack};
  long long start = 1;
  long long end = maxHp;
  long long ans = maxHp+1;
  while(start<=end){
    long long mid = (start+end)/2;
    me.cur = mid;
    me.max = mid;
    bool killDragon;
    killDragon = train(me, room_info);
  
    if(killDragon){  
    ans = min(ans, mid);
    end = mid - 1;
    }else{
    start = mid + 1;
    };
  };
  
  cout << ans << "\n";
  
  return 0;
}